### PR TITLE
Fix run slaves joining started runs

### LINKED ISSUE #188: Run failed with no project
A run failed when a mult user consort has one member who joined but had not created a project. Creating the project did _not_ jumpstart the run :-(

### PR DESCRIPTION
**Pull request by rssk**
Removes the check for a currently active local run, this allows 2
things:

1. users who have already started to rejoin
2. users who have not started to join

Also this does *not* state clear the local result db, otherwise the
local resulst are not available on the local runner

Closes #188

### CODE REVIEW

File: packages/coinstac-ui/package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "commander": "^2.9.0",
     "convict": "^1.4.0",
     "deep-equal": "^1.0.1",
+    "electron": "^1.5.1",
```
Comment (swashcap) [inline packages/coinstac-ui/package.json]: Why is electron a `dependency` and not a `devDependency`?

File: packages/coinstac-common/src/services/classes/db-registry.js [hunk 2]
```diff
@@ -188,14 +188,15 @@ class DBRegistry {
...
-     * memory.
+     * Ensure local (client) instances' databases are stored in
+     * memory. This excludes projects and local consortium results.
      *
      * {@link https://github.com/MRN-Code/coinstac/issues/155}
      */
     if (
       this.localStores &&
       this.localStores.indexOf('projects') > -1 &&
+      this.localStores.indexOf('local-consortium') > -1 &&
```
Comment (swashcap) [inline packages/coinstac-common/src/services/classes/db-registry.js:199]: I foresee this causing issues as _local-consortium-xxx_ databases are “sync-out.” Local and remote databases could become out of sync, causing difficult state problems.

File: packages/coinstac-client-core/src/sub-api/computation-service.js [hunk 3]
```diff
@@ -194,17 +194,8 @@ class ComputationService extends ModelService {
...
-         * permits multiple simultaneous runs.
-         */
-        const { username } = auth.getUser();
-
-        // Determine whether the user has a doc with the run ID:
-        return !localDocs.find(({ _id }) => {
-          return _id.indexOf(runId) > -1 && _id.indexOf(username) > -1;
-        });
+        // The client is part of the consortia, and there is an active run
+        return true;
```
Comment (swashcap) [inline packages/coinstac-client-core/src/sub-api/computation-service.js]: 😐

Why do we want a user to always join a run?

**Comment by swashcap** [comment]
A little background by examining callsites:

1. `ComputationService#shouldJoinRun` (in _coinstac-client-core_) determines whether a client _should_ run based on whether the current user is in the consortium and the consortium has a run ID. It _did_ do some local document/database checks prior to MRN-Code/coinstac#204.
2. `bgServices.joinSlaveComputation` (in _coinstac-ui_) joins a computation run if:
    * The client has a project associated with the consortium
    * The consortium has an active run ID (stored on the `Consortium` model’s `activeRunId` property)
    * The result of `ComputationService#shouldJoinRun` is “yep”
3. `bgServices.joinSlaveComputation` is called in three other `bgServices` functions:
    * `bgServices.initPrivateBackgroundServices` (in _coinstac-ui_) fetches the user's consortia and runs `bgServices.joinSlaveComputation` on each of them. This is called once the user logs in.
    * `bgServices.addConsortiumComputationListener` (in _coinstac-ui_) attaches a listener to a consortium's associated `remote-consortium-xxx` database. Every time a document changes in the database the handler fires `bgServices.joinSlaveComputation` with the consortium.
    * `projects.addProject` action creator (in _coinstac-ui_) fires `bgServices.joinSlaveComputation` after a new project is first saved.

`ComputationService#shouldJoinRun` did the following checks:

* Ensure the target consortium has an active run ID
* Determines if the user has a `local-consortium-xxx` with that active run ID. If:
  * **Yep:** User already has a document, meaning _they’ve already joined the run_. Return `false`.
  * **Nope:** User doesn't have a document, meaning they _haven't yet contributed to the run_. Return `true`.

Always returning `true` from `ComputationService#shouldJoinRun` ignores the local computation result contract that we've established. I don't believe we can use this approach.

File: packages/coinstac-common/src/services/classes/db-registry.js (hunk 2 above)
Comment (rssk) [reply packages/coinstac-common/src/services/classes/db-registry.js:199]: The local res doc used on the remote ?

**Comment by rssk** [comment]
Not sure I agree, if we exclude that case, if you've started a run, and restart cstac, it'll never rejoin. This allows that.

File: packages/coinstac-common/src/services/classes/db-registry.js (hunk 2 above)
Comment (swashcap) [reply packages/coinstac-common/src/services/classes/db-registry.js:199]: Yep, `RemotePipelineRunnerPool` grabs the `local-consortium-*` database for runner instantiation: https://github.com/MRN-Code/coinstac/blob/941a7ea50801aaf7e3df0e604d3f27b4c6c14874/packages/coinstac-common/src/models/pipeline/runner/pool/remote-pipeline-runner-pool.js#L31-L63. This makes sense: the server needs access to clients' results (documents in `local-consortium-*` database) to provide them to computations. Also, clients are configured to sync `local-consortium-*` out: https://github.com/MRN-Code/coinstac/blob/941a7ea50801aaf7e3df0e604d3f27b4c6c14874/packages/coinstac-common/src/services/db-registry.js#L58.

File: packages/coinstac-ui/package.json (hunk 1 above)
Comment (rssk) [reply packages/coinstac-ui/package.json]: prob a mischievous `-s`

**Comment by swashcap** [comment]
I’m seeing a computation-related error:


end-of-wat

Steps to reproduce:

1. Start client 1. Create a consortium, set up a project.
2. Start client 2. Join the consortium.
3. Click “run” on client 1.
4. Create a project in client 2 and save.
5. Observe _errors_ on client 2.

_(This works on `master`.)_

Here’s the stack trace:

```
Error: Cannot read property 'endOfIteration' of undefined
    at JavascriptComputation.fn (/Users/creed/Projects/coinstac/packages/coinstac-computation-registry/node_modules/laplacian-noise-ridge-regression/src/index.js:68:87)
    at JavascriptComputation.run (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/computation/javascript-computation.js:20:33)
    at maybeIncrementStep.then.then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/pipeline.js:159:35)
From previous event:
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at runQueue.(anonymous function).runQueue.(anonymous function).then (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:563:23)
    at LocalPipelineRunnerPool.triggerRunner (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:561:6)
    at LocalPipelineRunnerPool.handleResultChange (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:313:10)
    at DBListener.listener.on (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:416:24)
    at Changes.DBListener.changes.pouchy.changes.on.info (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/db-listener.js:30:14)
    at tryCatchInChangeListener (/Users/creed/Projects/coinstac/packages/coinstac-common/node_modules/pouchdb-core/lib/index.js:44:10)
    at Object.Changes.opts.onChange (/Users/creed/Projects/coinstac/packages/coinstac-common/node_modules/pouchdb-core/lib/index.js:82:5)
    at Changes.<anonymous> (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/pouchdb-utils/lib/index.js:409:14)
    at tryCatchInChangeListener (/Users/creed/Projects/coinstac/packages/coinstac-common/node_modules/pouchdb-core/lib/index.js:44:10)
    at Object.Changes.opts.onChange (/Users/creed/Projects/coinstac/packages/coinstac-common/node_modules/pouchdb-core/lib/index.js:82:5)
    at onGetWinningDoc (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/pouchdb-adapter-leveldb-core/lib/index.js:1319:20)
    at onGetMetadata (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/pouchdb-adapter-leveldb-core/lib/index.js:1337:18)
    at /Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/pouchdb-adapter-leveldb-core/lib/index.js:1361:9
    at /Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/sublevel-pouchdb/lib/shell.js:87:12
    at /Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/sublevel-pouchdb/lib/nut.js:103:19
From previous event:
    at LocalPipelineRunnerPool.triggerRunner (/Users/creed/Projects/coinstac/packages/coinstac-common/src/models/pipeline/runner/pool/pipeline-runner-pool.js:561:6)
    at Promise.all.then (/Users/creed/Projects/coinstac/packages/coinstac-client-core/src/sub-api/computation-service.js:138:21)
    at ComputationService.doTriggerRunner (/Users/creed/Projects/coinstac/packages/coinstac-client-core/src/sub-api/computation-service.js:98:8)
    at ComputationService.joinRun (/Users/creed/Projects/coinstac/packages/coinstac-client-core/src/sub-api/computation-service.js:172:17)
    at callFunction (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:251:18)
    at EventEmitter.<anonymous> (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:366:5)
    at WebContents.<anonymous> (/Users/creed/Projects/coinstac/packages/coinstac-ui/node_modules/electron/d…ctron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:249:37)
```

**Comment by swashcap** [comment]
Regular run seems to work fine!

188-reg-run

**Comment by swashcap** [comment]
✅ Delayed runs (second user joins, first user clicks “run,” second user creates project) work! 🎉

**Comment by swashcap** [comment]
coldest-restarts

✅ Killing a client, restarting mid-run.

**Comment by rssk** [comment]
Thank god, this was such a pain

**Comment by swashcap** [comment]
✅ Runs where a member doesn’t initially have a project work.

I do see this UI warning:

```
Warning: Failed prop type: Invalid prop `values[1]` of type `number` supplied to `ConsortiumComputationInputsViewer`, expected an array.
    in ConsortiumComputationInputsViewer (created by Consortium)
    in Consortium (created by ConsortiumController)
    in ConsortiumController (created by Connect(ConsortiumController))
    in Connect(ConsortiumController) (created by RouterContext)
    in div (created by Dashboard)
    in div (created by Dashboard)
    in div (created by Dashboard)
    in Dashboard (created by Connect(Dashboard))
    in Connect(Dashboard) (created by RouterContext)
    in div (created by App)
    in App (created by Connect(App))
    in Connect(App) (created by RouterContext)
    in RouterContext (created by Router)
    in Router
    in Provider
```

Taking a quick peek.

**Update:** Bad prop validation. I'll fix in future PR.

File: packages/coinstac-client-core/src/sub-api/computation-service.js [hunk 4]
```diff
@@ -173,12 +173,65 @@ class ComputationService extends ModelService {
   }
 
   /**
+   * Allows the client to join a run for which they have no started document for.
+   * @param  {string} options.consortiumId Consortium the runId is from
+   * @param  {string} options.projectId    project to run on
+   * @param  {string} options.runId
+   * @return {Promise}
+   */
+  joinSlavedRun({ consortiumId, projectId, runId }) {
```
Comment (swashcap) [inline packages/coinstac-client-core/src/sub-api/computation-service.js:189]: This function body closely matches `ComputationService#doTriggerRunner`. DRY it out?

**Comment by swashcap** [comment]
Just did a multi-shot run through:

wat-happens

Things _work_, but the non-owner (right) gets a document update conflict after project save/run start. Shall I just merge and we can tackle the bug in another PR?

**Pull request merged**